modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/lifecycle/DefaultArtifactDisposalContext.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.lifecycle;
+
+import static java.lang.Thread.getAllStackTraces;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.runtime.core.internal.util.CompositeClassLoader;
+import org.mule.runtime.module.artifact.api.classloader.ArtifactClassLoader;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactDisposalContext;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactLifecycleListener;
+
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+
+/**
+ * Default implementation of {@link ArtifactDisposalContext}.
+ */
+public class DefaultArtifactDisposalContext implements ArtifactDisposalContext {
+
+  private static final Logger LOGGER = getLogger(DefaultArtifactDisposalContext.class);
+
+  private final ArtifactClassLoader artifactClassLoader;
+  private final ArtifactClassLoader extensionClassLoader;
+
+  /**
+   * Creates an {@link ArtifactDisposalContext} from the given {@code artifactClassLoader} and {@code extensionClassLoader} pair.
+   *
+   * @param artifactClassLoader  the {@link ArtifactClassLoader} of the artifact being disposed.
+   * @param extensionClassLoader the {@link ArtifactClassLoader} of the extension the {@link ArtifactLifecycleListener} is
+   *                             associated with.
+   */
+  public DefaultArtifactDisposalContext(ArtifactClassLoader artifactClassLoader, ArtifactClassLoader extensionClassLoader) {
+    this.artifactClassLoader = artifactClassLoader;
+    this.extensionClassLoader = extensionClassLoader;
+  }
+
+  @Override
+  public ClassLoader getExtensionClassLoader() {
+    return extensionClassLoader.getClassLoader();
+  }
+
+  @Override
+  public ClassLoader getArtifactClassLoader() {
+    return artifactClassLoader.getClassLoader();
+  }
+
+  @Override
+  public Stream<Thread> getArtifactOwnedThreads() {
+    try {
+      // This may be expensive because it needs to build all the stack traces which we don't actually need.
+      // However, other alternatives like Thread#enumerate would require us to traverse up to the root Thread Group in order
+      // to actually get all threads.
+      // Because this is only executed on disposal, we prefer readability to marginal performance gains.
```

Comment:
I meant we don't want to fall into premature optimization. But yes, you are right in that I was only considering CPU time. The memory pressure from this could be more of an issue. The problem is I don't think I can make a very representative benchmark of the best/avg/worst case scenarios because I don't know how those apps would look like.

Taking that into consideration maybe it is actually worth it to be safer on the memory usage side and add the complexity needed to use `Thread#enumerate` over all the `ThreadGroup` hierarchy.

Would you still think we should spend the time creating the benchmarks for this?